src/main/scala/org/camunda/feel/impl/parser/FeelParser.scala

```diff
@@ -118,7 +119,29 @@ object FeelParser {
 
   // 2 a)
   private def textualExpression[_: P]: P[Exp] =
-    P(functionDefinition | forExpression | ifExpression | quantifiedExpression | expression2)
+    P(expressionInBrackets | functionDefinition | forExpression | ifExpression | quantifiedExpression | expression2)
+
+  private def expressionInBrackets[_: P]: P[Exp] =
+    P("(" ~ expression ~ ")").flatMap(continuation)
+
+  // optimize parsing the expression from left to right by using `flapMap()`
+  // `flatMap()` takes the parsed (left) part of the expression and continues with the (right) part
+  // it does backtracking if `flatMap()` is not successful
+  private def continuation[_: P](base: Exp): P[Exp] = P(
+    continuationExpression(base).flatMap(continuation) |
+      End.map(_ => base)
+  )
+
+  private def continuationExpression[_: P](base: Exp): P[Exp] =
+    filterContinuation(base) | pathContinuation(base)
+
+  private def filterContinuation[_: P](base: Exp): P[Exp] =
+    P("[" ~ expression ~ "]")
+      .map(filter => Filter(base, filter))
```

Comment:
This seems kind of duplicate of `filteredExpression9`. Does this mean the other filterExpressions can be removed or replaced in the future?

Reply:
Yes. I plan to apply the new style to more places. If this is done then I'll remove the duplications :+1: 
Currently, I want to keep the changes small.